### ISSUE TITLE
WiFi state machine race conditions

### ISSUE THREAD

**Issue by downwith** [MicrochipTech/aws-iot-zero-touch-secure-provisioning-kit#5]
It appears there are callbacks in aws_wifi_state.c which are executed from ISRs and modify the state machine state without any synchronization methods.  I believe this will cause race conditions if the state machine thread is about to modify the state and a callback occurs which also modifies the state.  

e.g. The callback could handle a disconnect scenario and change the state to start the disconnect process, and then the thread could immediately change it back to a connected state if it was in the processes of updating the state.  This would likely cause the state machine to fall apart.

Shouldn't there be a synchronization between the callback and threads such that the state is updated in the thread and not in the callback?

**Comment by bryan-hunt** [comment]
The callbacks are made through the winc driver via m2m_wifi_handle_events which is executed in the context of the aws_wifi_task if you are referring to aws_wifi_callback in the aws_wifi_task.c file. The driver is polled so there are no ISRs in this context.

**Comment by downwith** [comment]
When I followed the WiFi callback, I saw the wifi init sets gpfAppWifiCb, which appears to be called in m2m_wifi_cb, which appears to be registered using hif_register_cb, which I saw was being handled by something which looked like an ISR (because it was called hif_isr).  Now that I dig further, I see that the ISRs are actually being handled by that event handler you mention.  Thanks for the correction!
